File: semana_3/solucion.py
import random


class PasswordGenerator:

    def __init__(self, longitud:int=8, con_mayusculas: bool = False, 
                con_numeros: bool=False, con_simbolos: bool = False):
        if longitud < 8 or longitud > 16:
            raise "La longitud debe ser de entre 8 y 16"
        self.longitud = longitud
        self.con_mayusculas = con_mayusculas
        self.con_numeros = con_numeros
        self.con_simbolos = con_simbolos

    def get_letras(self):
        letras = []
        for letra in 'abcdefghijklmnopqrstuvwxyz':
            letras.append(letra)
        return letras

    def get_simbolos(self):
        simbolos = []
        for simbolo in "!#$%&/()=?-.,@":
            simbolos.append(simbolo)
        return simbolos

    def get_numeros(self):
        numeros = []
        for numero in range(0,10,1):
            numeros.append(numero)
        return numeros

    def new_password(self):
        password = []
        longitud = 0
        letras = self.get_letras()
        random.shuffle(letras)
        composicion = [letras]
        if self.con_mayusculas:
            letras_mayusculas = [] 
            for letra in letras:
                letras_mayusculas.append(letra.upper())
            random.shuffle(letras_mayusculas)
            composicion.append(letras_mayusculas)
        
        if self.con_numeros:
            numeros = self.get_numeros()
            random.shuffle(numeros)
            composicion.append(numeros)

        if self.con_simbolos:
            simbolos = self.get_simbolos()
            random.shuffle(simbolos)
            composicion.append(simbolos)

        while longitud != self.longitud:
            for compositor in composicion:
                caracter = random.choice(compositor)
                password.append(str(caracter))
                longitud += 1
                if longitud == self.longitud:
                    break

        random.shuffle(password)
        password = "".join(password)
        return password 


new_password = PasswordGenerator(
    longitud=10,
    con_mayusculas=True,
    con_numeros=True,
    con_simbolos=True
).new_password()
print(new_password)

File: semana_3/solucion.dart
import 'dart:math';


class PasswordGenerator{
  int longitud;
  bool con_mayusculas;
  bool con_numeros;
  bool con_simbolos;

  PasswordGenerator(this.longitud, this.con_mayusculas, this.con_numeros, this.con_simbolos)
  : assert(
    longitud >= 8 && longitud <= 16,
    "Longitud debe ser un valor entre 8 y 16",
  );

  List obtener_letras(){
    var letrasList = [];
    const letras = "abcdefghijklmnopqrstuvwxyz";
    for (int posicion=0; posicion< letras.length; posicion++){
      letrasList.add(letras[posicion]);
    }
    return letrasList;
  }

  List obtener_simbolos(){
    List simbolosLista = [];
    const simbolos = "!#%&/()=?-.,@";
    for (int posicion=0; posicion < simbolos.length; posicion++){
      simbolosLista.add(simbolos[posicion]);
    }
    return simbolosLista;
  } 

  List obtener_numeros(){
    List numerosLista = [];
    for (int i=0; i < 10; i++){
      numerosLista.add(i.toString());
    }
    return numerosLista;
  }

  String generar(){
    List passwordList = [];
    List letras = obtener_letras();
    List caracteresDisponibles = [];
    final letrasSoporte = obtener_letras();
    final _random = new Random();
  
    if (con_mayusculas){
      for (int posicion = 0; posicion < letrasSoporte.length; posicion++){
        letras.add(letrasSoporte[posicion].toUpperCase());
      }
    }
    caracteresDisponibles += letras;
    if (con_numeros){
      caracteresDisponibles += obtener_numeros();
    }
    if (con_simbolos){
      caracteresDisponibles += obtener_simbolos();
    }
    for (int i = 0; i < longitud; i++){
      var caracter = caracteresDisponibles[_random.nextInt(caracteresDisponibles.length)];
      passwordList.add(caracter);
    }
    String password = passwordList.join("");
    return password;
  }
}

void main(){
  var passwordGenerator = PasswordGenerator(15, true, true, true);
  final password = passwordGenerator.generar(); 
  print("Su nueva contraseña es: $password");
}
